File: backend/pkg/api/insight/handlers.go
package insight

import (
	"backend/pkg/api/dto"
	"backend/pkg/api/middleware"
	"backend/pkg/api/response"
	"backend/pkg/api/validation"
	"backend/pkg/core"
	coreInsight "backend/pkg/core/insight"
	"backend/pkg/core/ml"
	"backend/pkg/database"
	"backend/pkg/database/repository"

	"github.com/gofiber/fiber/v2"
	"github.com/google/uuid"
)

// getService creates and returns an insight service instance
func getService() *coreInsight.Service {
	db := database.DB()
	return coreInsight.NewService(
		repository.NewInsightRepository(db),
		repository.NewInsightTopicRepository(db),
		repository.NewUserInsightStatusRepository(db),
		repository.NewCrawledContentRepository(db),
		repository.NewContentTopicMatchRepository(db),
		ml.NewEmbeddingService(),
	)
}

// =============================================================================
// Insight Handlers
// =============================================================================

// ListInsights handles GET /insights
// @Summary List insights
// @Description Get a list of all insights with user-specific read/pinned status
// @Tags insights
// @Accept json
// @Produce json
// @Param page query int false "Page number" default(1)
// @Param limit query int false "Items per page" default(20)
// @Param topic_id query string false "Filter by topic ID"
// @Success 200 {object} response.SuccessResponse{data=object{insights=[]dto.InsightWithUserStatus,total=int64}}
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /insights [get]
func ListInsights(c *fiber.Ctx) error {
	userID, err := middleware.GetUserID(c)
	if err != nil {
		return response.Error(c, err)
	}

	page := c.QueryInt("page", 1)
	limit := c.QueryInt("limit", 20)
	topicIDStr := c.Query("topic_id")

	svc := getService()

	// For topic filtering, use legacy function then merge with user status
	if topicIDStr != "" {
		topicID, parseErr := uuid.Parse(topicIDStr)
		if parseErr != nil {
			return response.Error(c, core.InvalidInputError("Invalid topic ID"))
		}
		insights, total, err := svc.ListInsightsByTopic(c.Context(), topicID, page, limit)
		if err != nil {
			return response.Error(c, err)
		}
		return response.Success(c, fiber.Map{
			"insights": insights,
			"total":    total,
			"page":     page,
			"limit":    limit,
		})
	}

	// Get insights with user-specific status
	insights, total, err := svc.ListInsightsWithUserStatus(c.Context(), userID, page, limit)
	if err != nil {
		return response.Error(c, err)
	}

	return response.Success(c, fiber.Map{
		"insights": insights,
		"total":    total,
		"page":     page,
		"limit":    limit,
	})
}

// GetInsight handles GET /insights/:id
// @Summary Get insight
// @Description Get an insight by ID with its source content
// @Tags insights
// @Accept json
// @Produce json
// @Param id path string true "Insight ID"
// @Success 200 {object} response.SuccessResponse{data=dto.InsightWithSources}
// @Failure 400 {object} response.SuccessResponse
// @Failure 404 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /insights/{id} [get]
func GetInsight(c *fiber.Ctx) error {
	idStr := c.Params("id")
	id, err := uuid.Parse(idStr)
	if err != nil {
		return response.Error(c, core.InvalidInputError("Invalid insight ID"))
	}

	svc := getService()
	insight, err := svc.GetInsightWithSources(c.Context(), id)
	if err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, insight)
}

// MarkInsightAsRead handles POST /insights/:id/read
// @Summary Mark insight as read
// @Description Mark an insight as read for the current user
// @Tags insights
// @Accept json
// @Produce json
// @Param id path string true "Insight ID"
// @Success 200 {object} response.SuccessResponse{data=object{success=boolean}}
// @Failure 400 {object} response.SuccessResponse
// @Failure 404 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /insights/{id}/read [post]
func MarkInsightAsRead(c *fiber.Ctx) error {
	userID, err := middleware.GetUserID(c)
	if err != nil {
		return response.Error(c, err)
	}

	idStr := c.Params("id")
	insightID, err := uuid.Parse(idStr)
	if err != nil {
		return response.Error(c, core.InvalidInputError("Invalid insight ID"))
	}

	svc := getService()
	if err := svc.MarkInsightAsReadForUser(c.Context(), userID, insightID); err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, fiber.Map{"success": true})
}

// ToggleInsightPinned handles POST /insights/:id/pin
// @Summary Toggle insight pinned status
// @Description Toggle the pinned status of an insight for the current user
// @Tags insights
// @Accept json
// @Produce json
// @Param id path string true "Insight ID"
// @Success 200 {object} response.SuccessResponse{data=object{success=boolean,is_pinned=boolean}}
// @Failure 400 {object} response.SuccessResponse
// @Failure 404 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /insights/{id}/pin [post]
func ToggleInsightPinned(c *fiber.Ctx) error {
	userID, err := middleware.GetUserID(c)
	if err != nil {
		return response.Error(c, err)
	}

	idStr := c.Params("id")
	insightID, err := uuid.Parse(idStr)
	if err != nil {
		return response.Error(c, core.InvalidInputError("Invalid insight ID"))
	}

	svc := getService()
	isPinned, err := svc.ToggleInsightPinnedForUser(c.Context(), userID, insightID)
	if err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, fiber.Map{"success": true, "is_pinned": isPinned})
}

// SearchInsights handles GET /insights/search
// @Summary Search insights
// @Description Search insights using semantic similarity
// @Tags insights
// @Accept json
// @Produce json
// @Param q query string true "Search query"
// @Param limit query int false "Max results" default(10)
// @Success 200 {object} response.SuccessResponse{data=[]dto.InsightResponse}
// @Failure 400 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /insights/search [get]
func SearchInsights(c *fiber.Ctx) error {
	query := c.Query("q")
	if query == "" {
		return response.Error(c, core.InvalidInputError("Search query required"))
	}

	limit := c.QueryInt("limit", 10)

	svc := getService()
	orgID := middleware.GetOrgID(c)
	if orgID != nil {
		insights, err := svc.SearchInsightsByOrg(c.Context(), *orgID, query, limit)
		if err != nil {
			return response.Error(c, err)
		}
		return response.Success(c, insights)
	}

	req := dto.InsightSearchRequest{
		Query: query,
		Limit: limit,
	}

	insights, err := svc.SearchInsights(c.Context(), req)
	if err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, insights)
}

// GetUnreadCount handles GET /insights/unread-count
// @Summary Get unread insight count
// @Description Get the count of unread insights for the current user
// @Tags insights
// @Accept json
// @Produce json
// @Success 200 {object} response.SuccessResponse{data=object{count=int64}}
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /insights/unread-count [get]
func GetUnreadCount(c *fiber.Ctx) error {
	userID, err := middleware.GetUserID(c)
	if err != nil {
		return response.Error(c, err)
	}

	svc := getService()
	count, err := svc.CountUnreadInsightsForUser(c.Context(), userID)
	if err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, fiber.Map{"count": count})
}

// DeleteInsight handles DELETE /insights/:id
// @Summary Delete insight
// @Description Delete an insight by ID
// @Tags insights
// @Accept json
// @Produce json
// @Param id path string true "Insight ID"
// @Success 200 {object} response.SuccessResponse{data=object{success=boolean}}
// @Failure 400 {object} response.SuccessResponse
// @Failure 404 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /insights/{id} [delete]
func DeleteInsight(c *fiber.Ctx) error {
	idStr := c.Params("id")
	id, err := uuid.Parse(idStr)
	if err != nil {
		return response.Error(c, core.InvalidInputError("Invalid insight ID"))
	}

	svc := getService()
	if err := svc.DeleteInsight(c.Context(), id); err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, fiber.Map{"success": true})
}

// =============================================================================
// Topic Handlers
// =============================================================================

// ListTopics handles GET /insights/topics
// @Summary List topics
// @Description Get a list of all insight topics
// @Tags insights
// @Accept json
// @Produce json
// @Success 200 {object} response.SuccessResponse{data=[]dto.InsightTopicResponse}
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /insights/topics [get]
func ListTopics(c *fiber.Ctx) error {
	orgID := middleware.GetOrgID(c)

	var topics []dto.InsightTopicResponse
	var err error

	svc := getService()
	if orgID != nil {
		topics, err = svc.ListTopics(c.Context(), *orgID)
	} else {
		topics, err = svc.ListAllTopics(c.Context())
	}

	if err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, topics)
}

// GetTopic handles GET /insights/topics/:id
// @Summary Get topic
// @Description Get a topic by ID
// @Tags insights
// @Accept json
// @Produce json
// @Param id path string true "Topic ID"
// @Success 200 {object} response.SuccessResponse{data=dto.InsightTopicResponse}
// @Failure 400 {object} response.SuccessResponse
// @Failure 404 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /insights/topics/{id} [get]
func GetTopic(c *fiber.Ctx) error {
	idStr := c.Params("id")
	id, err := uuid.Parse(idStr)
	if err != nil {
		return response.Error(c, core.InvalidInputError("Invalid topic ID"))
	}

	svc := getService()
	topic, err := svc.GetTopicByID(c.Context(), id)
	if err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, topic)
}

// CreateTopic handles POST /insights/topics
// @Summary Create topic
// @Description Create a new insight topic
// @Tags insights
// @Accept json
// @Produce json
// @Param request body dto.InsightTopicCreateRequest true "Topic details"
// @Success 201 {object} response.SuccessResponse{data=dto.InsightTopicResponse}
// @Failure 400 {object} response.SuccessResponse
// @Failure 401 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /insights/topics [post]
func CreateTopic(c *fiber.Ctx) error {
	var req dto.InsightTopicCreateRequest
	if err := c.BodyParser(&req); err != nil {
		return response.Error(c, core.InvalidInputError("Invalid request body"))
	}
	if err := validation.ValidateStruct(req); err != nil {
		return response.Error(c, err)
	}

	orgID := middleware.GetOrgID(c)

	svc := getService()
	topic, err := svc.CreateTopic(c.Context(), orgID, req)
	if err != nil {
		return response.Error(c, err)
	}
	return response.Created(c, topic)
}

// UpdateTopic handles PUT /insights/topics/:id
// @Summary Update topic
// @Description Update an existing topic
// @Tags insights
// @Accept json
// @Produce json
// @Param id path string true "Topic ID"
// @Param request body dto.InsightTopicUpdateRequest true "Topic update details"
// @Success 200 {object} response.SuccessResponse{data=dto.InsightTopicResponse}
// @Failure 400 {object} response.SuccessResponse
// @Failure 404 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /insights/topics/{id} [put]
func UpdateTopic(c *fiber.Ctx) error {
	idStr := c.Params("id")
	id, err := uuid.Parse(idStr)
	if err != nil {
		return response.Error(c, core.InvalidInputError("Invalid topic ID"))
	}

	var req dto.InsightTopicUpdateRequest
	if err := c.BodyParser(&req); err != nil {
		return response.Error(c, core.InvalidInputError("Invalid request body"))
	}

	svc := getService()
	topic, err := svc.UpdateTopic(c.Context(), id, req)
	if err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, topic)
}

// DeleteTopic handles DELETE /insights/topics/:id
// @Summary Delete topic
// @Description Delete a topic by ID
// @Tags insights
// @Accept json
// @Produce json
// @Param id path string true "Topic ID"
// @Success 200 {object} response.SuccessResponse{data=object{success=boolean}}
// @Failure 400 {object} response.SuccessResponse
// @Failure 404 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /insights/topics/{id} [delete]
func DeleteTopic(c *fiber.Ctx) error {
	idStr := c.Params("id")
	id, err := uuid.Parse(idStr)
	if err != nil {
		return response.Error(c, core.InvalidInputError("Invalid topic ID"))
	}

	svc := getService()
	if err := svc.DeleteTopic(c.Context(), id); err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, fiber.Map{"success": true})
}

// =============================================================================
// Crawled Content Handlers
// =============================================================================

// SearchCrawledContent handles GET /insights/content/search
// @Summary Search crawled content
// @Description Search crawled content using semantic similarity
// @Tags insights
// @Accept json
// @Produce json
// @Param q query string true "Search query"
// @Param limit query int false "Max results" default(10)
// @Success 200 {object} response.SuccessResponse{data=[]dto.CrawledContentResponse}
// @Failure 400 {object} response.SuccessResponse
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /insights/content/search [get]
func SearchCrawledContent(c *fiber.Ctx) error {
	query := c.Query("q")
	if query == "" {
		return response.Error(c, core.InvalidInputError("Search query required"))
	}

	limit := c.QueryInt("limit", 10)

	svc := getService()
	orgID := middleware.GetOrgID(c)
	if orgID != nil {
		contents, err := svc.SearchCrawledContentByOrg(c.Context(), *orgID, query, limit)
		if err != nil {
			return response.Error(c, err)
		}
		return response.Success(c, contents)
	}

	contents, err := svc.SearchCrawledContent(c.Context(), query, limit)
	if err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, contents)
}

// GetRecentCrawledContent handles GET /insights/content/recent
// @Summary Get recent crawled content
// @Description Get recently crawled content for the organization
// @Tags insights
// @Accept json
// @Produce json
// @Param limit query int false "Max results" default(20)
// @Success 200 {object} response.SuccessResponse{data=[]dto.CrawledContentResponse}
// @Failure 500 {object} response.SuccessResponse
// @Security BearerAuth
// @Router /insights/content/recent [get]
func GetRecentCrawledContent(c *fiber.Ctx) error {
	orgID := middleware.GetOrgID(c)
	if orgID == nil {
		return response.Error(c, core.UnauthorizedError("Organization required"))
	}

	limit := c.QueryInt("limit", 20)

	svc := getService()
	contents, err := svc.GetRecentCrawledContent(c.Context(), *orgID, limit)
	if err != nil {
		return response.Error(c, err)
	}
	return response.Success(c, contents)
}

File: backend/src/api/insight/handlers.rs
use axum::{
    Json,
    extract::{Path, Query, State, rejection::JsonRejection},
    http::StatusCode,
};
use uuid::Uuid;

use crate::{
    api::{auth::AuthenticatedAccount, response::SuccessResponse},
    core::insight::InsightSearchRequest,
    error::AppError,
};

use super::{
    dto::{
        CountResponse, CrawledContentResponse, InsightListQuery, InsightListResponse,
        InsightResponse, InsightTopicCreateRequest, InsightTopicResponse,
        InsightTopicUpdateRequest, InsightWithSources, InsightWithUserStatus, PinResponse,
        RecentQuery, SearchQuery, SuccessFlag,
    },
    error::InsightApiError,
    state::InsightState,
};

type ApiResult<T> = Result<Json<SuccessResponse<T>>, InsightApiError>;

#[utoipa::path(
    get,
    path = "/insights",
    params(InsightListQuery),
    responses(
        (status = 200, body = SuccessResponse<InsightListResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "insights",
    operation_id = "listInsights"
)]
pub async fn list_insights(
    authenticated: AuthenticatedAccount,
    State(state): State<InsightState>,
    Query(query): Query<InsightListQuery>,
) -> ApiResult<InsightListResponse> {
    let user_id = authenticated.into_inner().into_inner();
    let page = query.page.unwrap_or(1);
    let limit = query.limit.unwrap_or(20);
    let (values, total) = if let Some(topic_id) = query.topic_id {
        let topic_id = Uuid::parse_str(&topic_id)
            .map_err(|_| AppError::InvalidInput("Invalid topic ID".to_owned()))?;
        let (values, total) = state
            .service()
            .list_insights_by_topic(topic_id, page, limit)
            .await?;
        (
            values
                .into_iter()
                .map(InsightWithUserStatus::from)
                .collect(),
            total,
        )
    } else {
        let (values, total) = state
            .service()
            .list_insights_with_user_status(user_id, page, limit)
            .await?;
        (values.into_iter().map(Into::into).collect(), total)
    };
    Ok(Json(SuccessResponse::new(InsightListResponse {
        insights: values,
        total,
        page,
        limit,
    })))
}

#[utoipa::path(
    get,
    path = "/insights/search",
    params(SearchQuery),
    responses(
        (status = 200, body = SuccessResponse<Vec<InsightResponse>>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope),
        (status = 502, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "insights",
    operation_id = "searchInsights"
)]
pub async fn search_insights(
    authenticated: AuthenticatedAccount,
    State(state): State<InsightState>,
    Query(query): Query<SearchQuery>,
) -> ApiResult<Vec<InsightResponse>> {
    if query.q.is_empty() {
        return Err(AppError::InvalidInput("Search query required".to_owned()).into());
    }
    let limit = query.limit.unwrap_or(10);
    let organization_id = state.organization_id(authenticated.into_inner()).await;
    let values = match organization_id {
        Some(id) => {
            state
                .service()
                .search_insights_by_org(id, &query.q, limit)
                .await?
        }
        None => {
            state
                .service()
                .search_insights(InsightSearchRequest {
                    query: query.q,
                    topic_id: None,
                    limit,
                    is_unread: None,
                })
                .await?
        }
    };
    Ok(Json(SuccessResponse::new(
        values.into_iter().map(Into::into).collect(),
    )))
}

#[utoipa::path(
    get,
    path = "/insights/unread-count",
    responses(
        (status = 200, body = SuccessResponse<CountResponse>),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "insights",
    operation_id = "getUnreadInsightCount"
)]
pub async fn get_unread_count(
    authenticated: AuthenticatedAccount,
    State(state): State<InsightState>,
) -> ApiResult<CountResponse> {
    let count = state
        .service()
        .count_unread_insights_for_user(authenticated.into_inner().into_inner())
        .await?;
    Ok(Json(SuccessResponse::new(CountResponse { count })))
}

#[utoipa::path(
    get,
    path = "/insights/topics",
    responses(
        (status = 200, body = SuccessResponse<Vec<InsightTopicResponse>>),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "insights",
    operation_id = "listInsightTopics"
)]
pub async fn list_topics(
    authenticated: AuthenticatedAccount,
    State(state): State<InsightState>,
) -> ApiResult<Vec<InsightTopicResponse>> {
    let organization_id = state.organization_id(authenticated.into_inner()).await;
    let values = match organization_id {
        Some(id) => state.service().list_topics(id).await?,
        None => state.service().list_all_topics().await?,
    };
    Ok(Json(SuccessResponse::new(
        values.into_iter().map(Into::into).collect(),
    )))
}

#[utoipa::path(
    post,
    path = "/insights/topics",
    request_body = InsightTopicCreateRequest,
    responses(
        (status = 201, body = SuccessResponse<InsightTopicResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope),
        (status = 502, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "insights",
    operation_id = "createInsightTopic"
)]
pub async fn create_topic(
    authenticated: AuthenticatedAccount,
    State(state): State<InsightState>,
    body: Result<Json<InsightTopicCreateRequest>, JsonRejection>,
) -> Result<(StatusCode, Json<SuccessResponse<InsightTopicResponse>>), InsightApiError> {
    let Json(request) = body.map_err(invalid_request_body)?;
    let organization_id = state.organization_id(authenticated.into_inner()).await;
    let value = state
        .service()
        .create_topic(
            organization_id,
            request.validate().map_err(InsightApiError::validation)?,
        )
        .await?;
    Ok((
        StatusCode::CREATED,
        Json(SuccessResponse::new(value.into())),
    ))
}

#[utoipa::path(
    get,
    path = "/insights/topics/{id}",
    params(("id" = Uuid, Path)),
    responses(
        (status = 200, body = SuccessResponse<InsightTopicResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 404, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "insights",
    operation_id = "getInsightTopic"
)]
pub async fn get_topic(
    _authenticated: AuthenticatedAccount,
    State(state): State<InsightState>,
    Path(id): Path<String>,
) -> ApiResult<InsightTopicResponse> {
    let id = parse_topic_id(&id)?;
    Ok(Json(SuccessResponse::new(
        state.service().get_topic_by_id(id).await?.into(),
    )))
}

#[utoipa::path(
    put,
    path = "/insights/topics/{id}",
    params(("id" = Uuid, Path)),
    request_body = InsightTopicUpdateRequest,
    responses(
        (status = 200, body = SuccessResponse<InsightTopicResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 404, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope),
        (status = 502, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "insights",
    operation_id = "updateInsightTopic"
)]
pub async fn update_topic(
    _authenticated: AuthenticatedAccount,
    State(state): State<InsightState>,
    Path(id): Path<String>,
    body: Result<Json<InsightTopicUpdateRequest>, JsonRejection>,
) -> ApiResult<InsightTopicResponse> {
    let id = parse_topic_id(&id)?;
    let Json(request) = body.map_err(invalid_request_body)?;
    Ok(Json(SuccessResponse::new(
        state
            .service()
            .update_topic(id, request.into())
            .await?
            .into(),
    )))
}

#[utoipa::path(
    delete,
    path = "/insights/topics/{id}",
    params(("id" = Uuid, Path)),
    responses(
        (status = 200, body = SuccessResponse<SuccessFlag>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 404, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "insights",
    operation_id = "deleteInsightTopic"
)]
pub async fn delete_topic(
    _authenticated: AuthenticatedAccount,
    State(state): State<InsightState>,
    Path(id): Path<String>,
) -> ApiResult<SuccessFlag> {
    let id = parse_topic_id(&id)?;
    state.service().delete_topic(id).await?;
    Ok(Json(SuccessResponse::new(SuccessFlag { success: true })))
}

#[utoipa::path(
    get,
    path = "/insights/content/search",
    params(SearchQuery),
    responses(
        (status = 200, body = SuccessResponse<Vec<CrawledContentResponse>>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope),
        (status = 502, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "insights",
    operation_id = "searchInsightContent"
)]
pub async fn search_crawled_content(
    authenticated: AuthenticatedAccount,
    State(state): State<InsightState>,
    Query(query): Query<SearchQuery>,
) -> ApiResult<Vec<CrawledContentResponse>> {
    if query.q.is_empty() {
        return Err(AppError::InvalidInput("Search query required".to_owned()).into());
    }
    let limit = query.limit.unwrap_or(10);
    let organization_id = state.organization_id(authenticated.into_inner()).await;
    let values = match organization_id {
        Some(id) => {
            state
                .service()
                .search_crawled_content_by_org(id, &query.q, limit)
                .await?
        }
        None => {
            state
                .service()
                .search_crawled_content(&query.q, limit)
                .await?
        }
    };
    Ok(Json(SuccessResponse::new(
        values.into_iter().map(Into::into).collect(),
    )))
}

#[utoipa::path(
    get,
    path = "/insights/content/recent",
    params(RecentQuery),
    responses(
        (status = 200, body = SuccessResponse<Vec<CrawledContentResponse>>),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "insights",
    operation_id = "getRecentInsightContent"
)]
pub async fn get_recent_crawled_content(
    authenticated: AuthenticatedAccount,
    State(state): State<InsightState>,
    Query(query): Query<RecentQuery>,
) -> ApiResult<Vec<CrawledContentResponse>> {
    let organization_id = state
        .organization_id(authenticated.into_inner())
        .await
        .ok_or(AppError::Unauthorized)?;
    let values = state
        .service()
        .get_recent_crawled_content(organization_id, query.limit.unwrap_or(20))
        .await?;
    Ok(Json(SuccessResponse::new(
        values.into_iter().map(Into::into).collect(),
    )))
}

#[utoipa::path(
    get,
    path = "/insights/{id}",
    params(("id" = Uuid, Path)),
    responses(
        (status = 200, body = SuccessResponse<InsightWithSources>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 404, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "insights",
    operation_id = "getInsight"
)]
pub async fn get_insight(
    _authenticated: AuthenticatedAccount,
    State(state): State<InsightState>,
    Path(id): Path<String>,
) -> ApiResult<InsightWithSources> {
    let id = parse_insight_id(&id)?;
    Ok(Json(SuccessResponse::new(
        state.service().get_insight_with_sources(id).await?.into(),
    )))
}

#[utoipa::path(
    delete,
    path = "/insights/{id}",
    params(("id" = Uuid, Path)),
    responses(
        (status = 200, body = SuccessResponse<SuccessFlag>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 404, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "insights",
    operation_id = "deleteInsight"
)]
pub async fn delete_insight(
    _authenticated: AuthenticatedAccount,
    State(state): State<InsightState>,
    Path(id): Path<String>,
) -> ApiResult<SuccessFlag> {
    let id = parse_insight_id(&id)?;
    state.service().delete_insight(id).await?;
    Ok(Json(SuccessResponse::new(SuccessFlag { success: true })))
}

#[utoipa::path(
    post,
    path = "/insights/{id}/read",
    params(("id" = Uuid, Path)),
    responses(
        (status = 200, body = SuccessResponse<SuccessFlag>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 404, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "insights",
    operation_id = "markInsightRead"
)]
pub async fn mark_insight_as_read(
    authenticated: AuthenticatedAccount,
    State(state): State<InsightState>,
    Path(id): Path<String>,
) -> ApiResult<SuccessFlag> {
    let id = parse_insight_id(&id)?;
    state
        .service()
        .mark_insight_as_read_for_user(authenticated.into_inner().into_inner(), id)
        .await?;
    Ok(Json(SuccessResponse::new(SuccessFlag { success: true })))
}

#[utoipa::path(
    post,
    path = "/insights/{id}/pin",
    params(("id" = Uuid, Path)),
    responses(
        (status = 200, body = SuccessResponse<PinResponse>),
        (status = 400, body = crate::error::ErrorEnvelope),
        (status = 401, body = crate::error::ErrorEnvelope),
        (status = 404, body = crate::error::ErrorEnvelope),
        (status = 500, body = crate::error::ErrorEnvelope)
    ),
    security(("bearerAuth" = [])),
    tag = "insights",
    operation_id = "toggleInsightPinned"
)]
pub async fn toggle_insight_pinned(
    authenticated: AuthenticatedAccount,
    State(state): State<InsightState>,
    Path(id): Path<String>,
) -> ApiResult<PinResponse> {
    let id = parse_insight_id(&id)?;
    let is_pinned = state
        .service()
        .toggle_insight_pinned_for_user(authenticated.into_inner().into_inner(), id)
        .await?;
    Ok(Json(SuccessResponse::new(PinResponse {
        success: true,
        is_pinned,
    })))
}

fn parse_topic_id(value: &str) -> Result<Uuid, InsightApiError> {
    Uuid::parse_str(value).map_err(|_| AppError::InvalidInput("Invalid topic ID".to_owned()).into())
}

fn parse_insight_id(value: &str) -> Result<Uuid, InsightApiError> {
    Uuid::parse_str(value)
        .map_err(|_| AppError::InvalidInput("Invalid insight ID".to_owned()).into())
}

fn invalid_request_body(_: JsonRejection) -> InsightApiError {
    AppError::InvalidInput("Invalid request body".to_owned()).into()
}
